[Charmaine Ho] iP

charmainehly opened the pull request:
### DukePro
> “Your mind is for having ideas, not holding them.” 
> – David Allen

- based on user text input
- easy to understand
- ~~fast~~ _SUPER DUPER UBER_ fast

Here is what **YOU** need to do:
1. download the jar file from [here](https://github.com/charmainehly/ip/releases/tag/A-Jar)
2. run the jar file
3. start typing your task commands! 😄 

Here are the commands you can do:
- [X] Add tasks
- [X] Delete tasks
- [X] Search for tasks
- [X] Mark/Unmark tasks


If you Java programmer, you can use it to practice Java too. Here's part of the `main` method:
```ruby
    public static void main(String[] args) {
        new Duke("taskHistory.txt").run();
    }
```

Commit:
Add Gradle support

charmainehly committed:
Level-1 v1.0

charmainehly committed:
Level-2 v1.0

charmainehly committed:
Level-3 v1.0

charmainehly committed:
Level-4 v1.0 n A-TextUiTesting

charmainehly committed:
(add tag) A-TextUiTesting

charmainehly committed:
Level-5 v1.0

charmainehly committed:
Level-6 v1.0

charmainehly committed:
CREATE: create level-7 save (v1)

charmainehly committed:
CREATE: create level-8 dates and times (v1)

charmainehly committed:
Merge branch 'branch-Level-7'

charmainehly committed:
Merge branch 'branch-Level-8'

charmainehly committed:
UPDATE: add tags for Level-7 and Level-8

charmainehly committed:
UPDATE: add tags for Level-7 due to previous error

charmainehly committed:
Extract Ui, Storage, Parser and TaskList classes

The current driver in the main method lumps everything that needs to
be done, there is no good OOP practice in place.

Let's create separate classes to encapsulate the different methods that
need to be called.

Separating the different methods into the relevant classes help to make
the code more readable and will help with scalability.

charmainehly committed:
Add JUnit tests to test behaviour of toString()

Currently no JUnit tests to test for correct behaviour of methods.

Errors in toString() methods can lead to the wrong output being printed.

Let's add separate Test classes to check if toString() prints the
correct output for a specific instance of the class object.

charmainehly committed:
Create a jar file for the main module

tanyjnaaman commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,7 @@
+package duke;
```
Perhaps consider organizing exceptions into their own package, since it would help developers easily understand the code base when all classes have been grouped into packages!

tanyjnaaman replied:
Perhaps consider organizing exceptions into their own package, since it would help developers easily understand the code base when all classes have been grouped into packages!

tanyjnaaman commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,29 @@
+package duke;
+
+public class Parser {
```
Consider writing javadoc method headers to conform to the coding standard of public methods!

charmainehly replied:
Thanks for the comment! I have included JavaDoc comments in a commit further downstream! :)

tanyjnaaman commented on src/main/java/duke/Todo.java:
```diff
@@ -0,0 +1,17 @@
+package duke;
+
+public class Todo extends Task {
```
Consider naming in Pascal case to conform to the coding standard!

charmainehly replied:
Thanks for reminding on this! I've edited the class name in a commit further downstream :)

tanyjnaaman commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,145 @@
+        }
+    }
+
+    public void run() {
```
Adding a method description to conform to the coding standard would be helpful, perhaps consider describing the logic of the method too, since it's very complex.

tanyjnaaman commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,39 @@
+
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
Well done for remembering to name boolean variables appropriately!

tanyjnaaman reviewed (commented):
Overall, everything looks good in terms of the coding standard, the main area to focus on would only be to add JavaDoc comments! Keep it up, pal.

charmainehly committed:
Modify accepted input conventions in Parser class

Lack of standard accepted user input causes inconsistent behaviour when
user enters different inputs.

Difficult to troubleshoot errors when dealing with user inputs or catch
errors to throw new DukeException. Need to standardise what user input
is accepted when adding Tasks for Deadline, Event and Todo so that
expected behaviour of Duke chatbot is more intuitive and standardised.

Let's modify how the user input is parsed in by changing the
String.split() method arguments. The description and date/time for
Deadline and Event will be split by " - by: " and " - at: "
respectively. The date/time will be accepted in the format dd-MM-yyyy,
and will be read from the local taskHistory.txt file in the format MMM
dd yyyy.

Adjusting the acceptable user input format will prevent special
characters from affecting the behaviour of the String.split() method.
Standardising how the description and date/time is separated will allow
for more intuitive and standardised use of the Duke chatbot.

charmainehly committed:
Merge branch 'branch-A-JavaDoc' for latest edits

charmainehly committed:
Update the jar file built by IntelliJ

Modification to code in previous commit requires a new Jar file to be
built.

Let's re-build a new Jar file with the main Duke class.

charmainehly committed:
Add JavaDoc comments to public methods and classes

charmainehly committed:
Merge branch 'branch-A-JavaDoc'

charmainehly committed:
Modify naming to comply with coding standards

charmainehly committed:
Merge branch 'branch-A-CodingStandard'

Naming for some variables and classes do not comply with coding
standards.

Let's refactor and rename the variables and classes to ensure compliance
with coding standards.

charmainehly committed:
Create methods to find tasks with specific keyword

Current functionalities only allow printing of all tasks in the task list.

User may want to search and print only tasks containing a specific
keyword.

Let's create a new method in the TaskList class to filter the original
task list, to return a new task list with only tasks containing the
specified keyword.

charmainehly committed:
Merge branch 'branch-Level-9'

yumengtan commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,50 @@
+package duke;
```
Perhaps consider organizing some classes into their own package such like @tanyjnaaman has mentioned!

yumengtan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,176 @@
+     * Switches between cases based on the command passed in.
+     * Exits the running chat-bot when the 'bye' command is read from user input.
+     */
+    public void run() {
```
This method seems a little lengthy which might make it a little difficult to follow the flow. Perhaps a more concise method/abstraction methods could be employed!

charmainehly replied:
Yep, I'll definitely try to look into this for future revisions and commits :)

yumengtan reviewed (commented):
Overall, I think your code structure is very well organised! You have followed the Java Coding Standard closely and is clear and concise to understand. Overall, LGTM!

charmainehly committed:
Add parent directory to duke package 'seedu.duke'

charmainehly committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

charmainehly committed:
Add Gradle to project

charmainehly committed:
Merge branch 'branch-A-Gradle'

charmainehly committed:
Merge branch with Gradle plugin installed

Current project set up does not allow for automated running of tests.

Gradle plugin will enable building and running Duke. Subsequently can
also be set up to run unit tests.

Let's merge the branch that currently has the Gradle plugin set up and
installed into the main master branch.

charmainehly committed:
Add dependencies in build.gradle for gui set up

Denniszedead commented on src/main/java/seedu/duke/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDate by;
```
The ``by`` name is very vague. Perhaps you can give it a more specific name?

Denniszedead commented on src/main/java/seedu/duke/Event.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class Event extends Task {
+
+    protected LocalDate at;
```
Same here, if possible, perhaps you can give it a meaningful name?

Denniszedead commented on src/main/java/seedu/duke/Storage.java:
```diff
@@ -0,0 +1,65 @@
+public class Storage {
+
+    private String filePath;
+    private File f;
+    private Scanner s;
```
Since f and s are used beyond this scope, perhaps you can give it a meaningful name instead?

Denniszedead commented on src/main/java/seedu/duke/Storage.java:
```diff
@@ -0,0 +1,65 @@
+     * @return Returns an arraylist containing each line from the file as a separate string.
+     */
+    public ArrayList<String> load() {
+        ArrayList<String> tasksRead = new ArrayList<>();
```
Should it be named tasksReads instead?

Denniszedead commented on src/main/java/seedu/duke/TaskList.java:
```diff
@@ -0,0 +1,129 @@
+     * @return Returns a new task list object with the specified keyword.
+     */
+    public TaskList findTasks(String keyword) throws DukeException {
+        ArrayList<String> result = new ArrayList<String>();
```
Perhaps you can change this to ``results``

Denniszedead commented on src/main/java/seedu/duke/Ui.java:
```diff
@@ -0,0 +1,52 @@
+ * Contains methods to print out the welcome and exit lines, and to get the next user input.
+ */
+public class Ui {
+
```
Perhaps, there is no need to leave a space here.

Denniszedead commented on src/main/java/seedu/duke/Storage.java:
```diff
@@ -0,0 +1,65 @@
+     */
+    public Storage(String filePath) {
+        this.filePath = filePath;
+        this.f = new File(filePath);
```
Perhaps you can remove ``this``?

Denniszedead reviewed (commented):
Nice job 👍 !

I like that most of your code follow the coding standards! 

I observed that you did not use the words ``this`` to refer to object variables. That is good as to reduce unnecessary bugs! However, you might check your code in ``Storage.java`` again as there are a few usage of ``this`` on variable names not overshadowed by the parameters. 

All the best in CS2103!

charmainehly committed:
Set up basic functionality for GUI

Current set up only allowed CLI from the user, and System.out.println.

User may want to utilise a more user friendly and visual user interface.

Let's create the GUI to take in user inputs, and print out a generic
'Hi' response.

charmainehly committed:
Fix GUI to be functional for all commands

Current functionalities for GUI only allow a generic fixed response 'hi'
to any user input.

User would need the GUI to be able to respond with relevant messages
based on the commands input by the user.

Let's implement the run method of the Duke class to be able to return a
string of a specific format based on the command entered by the user. To
do this, we make the following main modifications:
    1. Create a command package to store all the different command
       classes.
    2. Create individual classes for each command type to implement the
       run() method to return a string of the required format - each of
       these command classes inherits from the generic Command class.
    3. Delete the Ui class which is no longer needed to interact with
       the user interface.

charmainehly committed:
Merge branch 'branch-Level-10'

charmainehly committed:
Create a new fat Jar file

There is currently no fat Jar file created.

Users on other computers will not be able to run this application
without local dependencies already installed.

Let's create a new fat Jar file using Gradle's ShadowJar.

The release of this fat Jar file will allow other users to run the
application independently on their own computers.

charmainehly committed:
Modify code abstraction to improve code quality

Duke class has a long getResponse() method which demonstrates poor code
quality. The Command subclasses are not utilising proper
implementations of inherited abstract classes.

Poor code quality and breach of coding standards can result in poor
code readability and also give rise to higher chance of errors.

Let's
* abstract away lower level code from the getResponse() method in the
  Duke class and implement the functionalities in the Parser class
* standardise the run() method parameters in the Command abstract
  subclasses and implement constructors
* create generic getCommandAction() and getCommandDetails() methods in
  the Parser class to handle the creation of new Command objects

Making use of abstraction and creating generic methods helps to reduce
repeated code and address the problems of code quality and standards.

Note that this commit does not refine and address other areas of
coding style.

charmainehly committed:
Abstract code to variables to reduce arrow-head style in TaskList class

charmainehly committed:
Merge pull request #2 from charmainehly/branch-A-CodeQuality

Branch a code quality

charmainehly committed:
Add assertions to classes in command, duke and task packages

There is no current use of assertions throughout the methods.

There are no complementary checks on the program state during runtime
apart from the use of exceptions, which can miss out on potential bugs.

Let's add assertions to the methods with parameter arguments needed to
check for null arguments and invalid integer values which should not be
passed in.

These assertions will help to verify the program state during run time
to ensure that the code is behaving as it is assumed to be.

charmainehly committed:
Merge pull request #3 from charmainehly/branch-A-Assertions

Add assertions to classes in command, duke and task packages

charmainehly committed:
Add method in TaskList class to search for any event clashes

There are no checks for date and time clashes when a user tries to
add an event task to the task list.

Multiple events held at the same date will be allowed to be added
to the existing task list, which does not showcase good 'task
management' functionality.

Let's add a findClashes method in the TaskList class to return true
if the date argument passed in matches any of the dates of existing
event tasks in the TaskList object.

Within this commit, bugs and errors in the Parser class are fixed
as well to fix the functionality of the task manager.

charmainehly committed:
Merge pull request #4 from charmainehly/branch-B-DetectAnomalies

Add method in TaskList class to search for any event clashes

charmainehly committed:
Modify code for parsing of input command and output display

The format of user input for adding event and deadline classes uses
" - at:" and " - by:" respectively. This format is standard across
the Command package class methods as well as in TaskList class.

We want to modify the output display to such that the dates are
placed within brackets. We want to modify the user input format
to accept " /at " and " /by ". This variation is subjective, to
improve the user display and input formatting.

Let's modify the following code:
* Add a toWrite() method in the Task class and subclasses to ensure
  the text to be written to the taskHistory.txt file is standardised
  (i.e. using " - at :" and " - by:")
* Modify the way user command is split in the command subclasses for
  AddEventCommand and AddDeadlineCommand to abide by the new variation
  (i.e. using " /at " and "/by "
* Modify the toString() method in the Task class and subclasses to
  ensure a standard format with dates placed within brackets for the
  Event and Deadline class

charmainehly committed:
Add description of application and summary of features to README.md

charmainehly committed:
Fix error and bugs in findTasks() method and related code

The findTasks() method does not work as expected  after previous
commit to vary the command input and user input formats.

Application is not able ot execute the findTasks() functionality
and attempts to enter "find [keyword]" results in an error.

Let's fix the code relating to the error:
* re-standardise the toWrite() string format in task subclasses
  (i.e. use " /by" and " /at" as the split regex parameter)
* re-standardise the constructor in TaskList class based on the
  modified toWrite() return format
* fix the code in the findTasks() method to ensure the right
  string formats is being returned and used

charmainehly committed:
Add Usage section for README.md file

There is currently no explanation of the command input usage.

Adding the Usage section will allow users to better understand
how to input commands to the application.

Lets' add the Usage section with each command action and other
required parameters required. Each command action has its own
example to showcase the expected outcome.

charmainehly committed:
Add try catch blocks for command package classes to throw exceptions

Exception handling for command package classes have not been implemented
to return and display the error message back to the user.

User will have no clarity and visibility as to what caused the exception
and which part of user input provided is wrong. By implementing these
code blocks, we can return the error message to the user.

Let's implement try catch blocks to catch IO and Array Index Out of
Bounds Exceptions. Throw new DukeException with the appropriate error
message to handle these caught exceptions.

charmainehly committed:
Refine code to add Javadocs and follow code quality and standards

Current code has some missing javadoc comments for method headers,
as well as some ambiguity with abiding by coding standard. There are
methods which showing very borderline arrowhead style of code.

Refine and update the code to abide by the above guidelines will
ensure standardisation and compliance.

Let's implement the following key changes:
* Run Gradle's checkstyle to identify checkstyle errors
* Refine code based on the errors highlighted by checkstyle
* Add javadoc headers to classes and comments to the public methods
* Refine some of the variable names to abide by coding standards
* Restructure code blocks to minimise arrowhead.

charmainehly committed:
Add catch block for DateTimeParseException in command subclasses

AddDeadlineCommand and AddEventCommand classes do not have a catch
block to deal with dates that have been input in the wrong format.

User will not be able to view the error message on the GUI window
if the dates entered for the command has not been parsed in properly.

Let's add a catch block in the two classes above, to deal with the
DateTimeParseException. The catch block will throw a new
DukeException, and and display a message to tell the user what is
the right date format to input.
a

charmainehly committed:
Modify command classes to handle exception for index out of bounds

TaskList class currently uses assertions to check that the task id
passed in is a valid index of the existing task list.

There is no message displayed back to the user in the case where the
task id is invalid.

Let's modify the method body in the TaskList class to remove the
assertions. Lets add a catch block in the command subclasses to handle
index array out of bounds exception to throw a new Duke exception which
will be handled.

charmainehly committed:
Update fat Jar file using Gradle ShadowJar

charmainehly committed:
Add override annotation for overridden method in Command child classes

Child classes of Command class currently do not have override
annotation for run() method.

Compiler is not able to check for overridden methods. This could
potentially lead to overlooked errors when implementing the methods in
child classes which are expected to override the method in the parent
class.

Let's add the overridden annotation to relevant methods in the child
classes.
